File: Bioinfo/src/main/java/codes/php/SW.php
<?php

function smith_score($seq1, $seq2, $match_penalty, $mismatch_penalty, $gap_penalty) {
    $len1 = strlen($seq1); 
    $len2 = strlen($seq2);

    $score_matrix = array_fill(0, $len1 + 1, array_fill(0, $len2 + 1, 0));

    $maxScore = 0;
    $max_pos = [0, 0];

    for ($i = 1; $i <= $len1; $i++) {
        for ($j = 1; $j <= $len2; $j++) {
            $score_diag = $score_matrix[$i - 1][$j - 1] + ((score($seq1[$i-1], $seq2[$j-1]) == 1) ? $match_penalty : $mismatch_penalty);
            $score_up = $score_matrix[$i - 1][$j] + $gap_penalty;
            $score_left = $score_matrix[$i][$j - 1] + $gap_penalty;
            $score_matrix[$i][$j] = max(0, $score_diag, $score_up, $score_left);

            if ($score_matrix[$i][$j] > $maxScore) {
                $maxScore = $score_matrix[$i][$j];
                $max_pos = [$i, $j];
            }
        }
    }
    return ['score' => $maxScore, 'score_pos' => $max_pos, 'matrix' => $score_matrix];
}

function score($char1, $char2) {
    return ($char1 == $char2) ? 1 : -1;
}

function smith_waterman($seq1, $seq2, $match_penalty, $mismatch_penalty, $gap_penalty) {

    $matrix = smith_score($seq1, $seq2, $match_penalty, $mismatch_penalty, $gap_penalty);
    list($i, $j) = $matrix['score_pos'];
    $score_matrix = $matrix['matrix'];
    $align1 = '';
    $align2 = '';
    $barrinha = '';
    $gaps = 0;

    while ($i > 0 && $j > 0 && $score_matrix[$i][$j] > 0) {
        if ($score_matrix[$i][$j] == $score_matrix[$i - 1][$j - 1] + score($seq1[$i - 1], $seq2[$j - 1])) {
            $align1 = $seq1[$i - 1] . $align1;
            $align2 = $seq2[$j - 1] . $align2;
            if (score($seq1[$i - 1], $seq2[$j - 1]) == 1) {
                $barrinha = '|' . $barrinha;
            } else {
                $barrinha = ':' . $barrinha;
            }                
            $i--;
            $j--;
        } elseif ($score_matrix[$i][$j] == $score_matrix[$i - 1][$j] + $gap_penalty) {
            $align1 = $seq1[$i - 1] . $align1;
            $align2 = '-' . $align2;
            $barrinha = '-' . $barrinha;
            $i--;
            $gaps++;
        } else {
            $align1 = '-' . $align1;
            $align2 = $seq2[$j - 1] . $align2;
            $barrinha = '-' . $barrinha;
            $j--;
            $gaps++;
        }
    }
    return ['score' => $matrix['score'], 'gaps' => $gaps, 'align1' => $align1, 'align2' => $align2, 'barrinha' => $barrinha];
}

// Verificar se o script foi chamado corretamente
// if ($argc < 3) {
//     die("Usage: php script.php <seq1> <seq2>\n");
// }

// Argumentos da linha de comando
$seq1 = $argv[1];
//$seq2 = $argv[2];
$seq2 = "TTTCGGCGAATTGAGAGAAATTAGGTGCGGTTTGTGTCTGAACCTTTTATCCTAGCGACGATTTTTTAAGGAAGTTGAATATGATCATCAAACCTAAAATTCGTGGATTTATCTGTACAACAACGCACCCAGTGGGTTGTGAAGCGAACGTAAAAGAACAAATTGCCTACACAAAAGCACAAGGTCCGATCAAAAACGCACCTAAGCGCGTGTTGGTTGTCGGATCGTCTAGCGGCTATGGTCTGTCATCACGCATCGCTGCGGCGTTTGGCGGTGGTGCGGCGACGATCGGCGTATTTTTCGAAAAGCCGGGCACTGACAAAAAACCAGGTACTGCGGGTTTCTACAATGCAGCAGCGTTTGACAAGCTAGCGCATGAAGCGGGCTTGTACGCAAAAAGCCTGAACGGCGATGCGTTCTCGAACGAAGCGAAGCAAAAAGCGATTGAGCTGATTAAGCAAGACCTCGGCCAGATTGATTTGGTGGTTTACTCATTGGCTTCTCCAGTGCGTAAAATGCCAGACACGGGTGAGCTAGTGCGCTCTGCACTAAAACCGATCGGCGAAACGTACACCTCTACCGCGGTAGATACCAATAAAGATGTGATCATTGAAGCCAGTGTTGAACCTGCGACCGAGCAAGAAATCGCTGACACTGTCACCGTGATGGGCGGTCAAGATTGGGAACTGTGGATCCAAGCACTGGAAGAGGCGGGTGTTCTTGCTGAAGGTTGCAAAACCGTGGCGTACAGCTACATCGGTACTGAATTGACTTGGCCAATCTACTGGGATGGCGCTTTAGGCCGTGCCAAGATGGACCTAGATCGCGCAGCGACAGCGCTGAACGAAAAGCTGGCAGCGAAAGGTGGTACCGCGAACGTTGCAGTTTTGAAATCAGTGGTGACTCAAGCAAGCTCTGCGATTCCTGTGATGCCGCTCTACATCGCGATGGTGTTCAAGAAGATGCGTGAACAGGGCGTGCATGAAGGCTGTATGGAGCAGATCTACCGCATGTTCAGTCAACGTCTGTACAAAGAAGATGGTTCAGCGCCGGAAGTGGATGATCACAATCGTCTGCGTTTGGATGACTGGGAACTGCGTGATGACATTCAGCAGCACTGCCGTGATCTGTGGCCACAAATCACTACAGAGAACCTGCGTGAGCTGACCGATTACGACATGTACAAAGAAGAGTTCATCAAGCTGTTTGGCTTTGGCATTGAAGGCATTGATTACGATGCTGACGTCAATCCAGAAGTCGAATTCGATGTGATTGATATCGAGTAAGAGAATTAACTCTTATCTTAAAAAGGCGCGTTATCGCGCCTTTTTTGTGTCCGGAGTACAGCATGAATACAGCAGGTTGC";

$start = microtime(true);
$result = smith_waterman($seq1, $seq2, 1, -1, -1);
$time = round((microtime(true) - $start),2);
$score = $result['score'];
$qtd_gap = $result['gaps'];
$barrinha = $result['barrinha'];
$percent = round(((100.0 * substr_count($barrinha, '|')) / strlen($barrinha)) * 100.0) / 100.0;

/*Percento:*/echo $percent . "\n";
/*Time:    */echo $time . "\n";
/*Score:   */echo $score . "\n";
/*Gap:     */echo $qtd_gap . "\n";

$text = "$time\n$score\n$qtd_gap\n$evalue";

    $arquivo = fopen("resultados_SW_php.txt", "w");
    if ($arquivo == false){
        die ('Não foi possível criar o arquivo.');
    } else{
        fwrite($arquivo, $text);
    }

?>

File: Bioinfo/src/main/java/codes/c/SW.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#define MATCH 1
#define MISMATCH -1
#define GAP -1
void swa(char*, char*);
int main(int argc, char *argv[]) {
  swa(argv[1], argv[2]);
  return 0;
}
void swa(char* seq1, char* seq2){
  clock_t ini = clock();
  int **mat = malloc((strlen(seq1)+1) * sizeof(int*));
  int **mat_dir = malloc((strlen(seq1)+1) * sizeof(int*));
  for(int i = 0; i<=strlen(seq1)+1; i++){
    mat[i] = malloc((strlen(seq2)+1) * sizeof(int));
    mat_dir[i] = malloc((strlen(seq2)+1) * sizeof(int*));
  }
  for (int i = 0; i <=strlen(seq1)+1; i++) {
    mat[i][0] = 0;
    mat_dir[i][0] = 0;
  }
  for(int i = 1; i <=strlen(seq2)+1; i++){
    mat[0][i] = 0;
    mat_dir[0][i] = 0;
  }
  int greater = 0;
  int greater_j = 0;
  int greater_i = 0;
  for(int i = 1; i<=strlen(seq1); i++){
    for(int j = 1; j<=strlen(seq2); j++){
      int x = 0;
      int y = 0;
      int z = 0;
      if(seq1[i-1] == seq2[j-1]){
        x = mat[i-1][j-1] + MATCH;
      }else{
        x = mat[i-1][j-1] + MISMATCH;
      }
      y = mat[i-1][j] + GAP;
      z = mat[i][j-1] + GAP;
      if(x >= y && x >= z && x>=0){
        mat[i][j] = x;
        mat_dir[i][j] = 3;
      }else if(y >= x && y >= z && y>=0){
        mat[i][j] = y;
        mat_dir[i][j] = 1;
      }else if(z > x && z > y && z>=0){
        mat[i][j] = z;
        mat_dir[i][j] = 2;
      }else{
        mat[i][j] = 0;
        mat_dir[i][j] = 0;
      }
      if(mat[i][j]>greater){
        greater = mat[i][j];
        greater_i = i;
        greater_j = j;
      }
    }
  }
  int col = greater_j;
  int lin = greater_i;
  int gaps = 0;
  int score = greater;
  while(mat[lin][col] != 0){
    if(mat_dir[lin][col] == 2){
      col--;
      gaps++;
    }else if(mat_dir[lin][col] == 1){
      lin--;
      gaps++;
    }else{
      col--;
      lin--;
    }
  }
  printf("%.2f\n", ((double)clock() - ini) / CLOCKS_PER_SEC);
  printf("%d\n", score);
  printf("%d\n", gaps);
  printf("62\n");
}